nixos/modules/services/matrix/draupnir.nix

```
@@ -0,0 +1,267 @@
+      description = lib.mdDoc ''
+        Whether Draupnir should log a lot more messages in the room,
+        mainly involves "all-OK" messages, and debugging messages for when draupnir checks bans in a room.
+      '';
+    };
+
+    dataPath = mkOption {
+      type = types.path;
+      default = "/var/lib/draupnir";
+      description = lib.mdDoc ''
+        The directory the bot should store various bits of information in.
+      '';
+    };
+
+    managementRoom = mkOption {
+      type = types.str;
+      default = "#moderators:example.org";
+      description = lib.mdDoc ''
+        The room ID or alias where moderators can use the bot's functionality.
+
+        The bot has no access controls, so anyone in this room can use the bot - secure this room!
+
+        Warning: When using a room alias, make sure the alias used is on the local homeserver!
+        This prevents an issue where the control room becomes undefined when the alias can't be resolved.
+      '';
+    };
+
+    protectedRooms = mkOption {
+      type = types.listOf types.str;
+      default = [ ];
+      example = literalExpression ''
+        [
+          "https://matrix.to/#/#yourroom:example.org"
+          "https://matrix.to/#/#anotherroom:example.org"
+        ]
+      '';
+      description = lib.mdDoc ''
+        A list of rooms to protect (matrix.to URLs).
+        These can also be configured interactively.
+      '';
+    };
+
+    settings = mkOption {
+      default = { };
+      type = (pkgs.formats.yaml { }).type;
+      example = literalExpression ''
+        {
+          autojoinOnlyIfManager = true;
+          automaticallyRedactForReasons = [ "spam" "advertising" ];
+        }
+      '';
+      description = lib.mdDoc ''
+        Additional settings (see [Draupnir's default configuration](https://github.com/the-draupnir-project/Draupnir/blob/main/config/default.yaml) for available settings).
+        These settings will override settings made by the module config.
+      '';
+    };
+  };
+
+  config = mkIf config.services.draupnir.enable {
+    assertions = [
+      {
+        assertion = !(cfg.pantalaimon.enable && cfg.pantalaimon.passwordFile == null);
+        message = "Specify pantalaimon.passwordFile";
+      }
+      {
+        assertion = !(cfg.pantalaimon.enable && cfg.accessTokenFile != null);
+        message = "Do not specify accessTokenFile when using pantalaimon";
+      }
+      {
+        assertion = !(!cfg.pantalaimon.enable && cfg.accessTokenFile == null);
+        message = "Specify accessTokenFile when not using pantalaimon";
+      }
+    ];
+
+    services.pantalaimon-headless.instances."draupnir" = mkIf cfg.pantalaimon.enable
+      {
+        homeserver = cfg.homeserverUrl;
+      } // cfg.pantalaimon.options;
+
+    systemd.services.draupnir = {
+      description = "Draupnir - a moderation tool for Matrix";
+      requires = [] ++ optionals (cfg.pantalaimon.enable) [
+        "pantalaimon-draupnir.service"
+      ];
```

```suggestion
      requires = optionals (cfg.pantalaimon.enable) [
        "pantalaimon-draupnir.service"
      ];
```
`lib.optional` will always return a list.